[KIP-71] Dynamic gas fee pricing mechanism

lwj010 opened the pull request:
KIP-71: A dynamic gas fee pricing mechanism based on network usage, that includes fixed-per-block network fee that is partially burned. 

Please refer to the https://forum.klaytn.com/t/en-new-transaction-fee-mechanism-dynamic-gas-price-proposal/4844/2 for discussion.

lwj010 committed:
KIP-17 proposal

lwj010 committed:
KIP-71

baerae-colin commented:
@lwj010 Could you please remove .DS_Store files?

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+---
+kip: 71
+title: Dynamic gas fee pricing mechanism
+author: Woojin Lee (jared) <jared.fi@krustuniverse.com>, Junghyun Colin Kim <colin.kim@groundx.xyz>
```
```suggestion
author: Woojin Lee (jared) <jared.fi@krustuniverse.com>, Junghyun Colin Kim <colin.klaytn@krustuniverse.com>
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+---
+kip: 71
+title: Dynamic gas fee pricing mechanism
```
```suggestion
title: Dynamic Gas Fee Pricing Mechanism
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+There is a base fee per gas in protocol, which can move up or down each block according to a formula which is a function of gas used in parent block and gas target. The algorithm results in the base fee per gas increasing when blocks are above the gas target, and decreasing when blocks are below the gas target. It repeats until base fee reachs lower bound fee or upper bound fee. The base fee per  ...
```
```suggestion
There is a base fee per gas in protocol, which can move up or down for each block according to a formula which is a function of the gas used in the parent block and the gas target. The algorithm results in the base fee per gas increasing when the gas used in the previous block is above the gas target, and decreasing when the gas used in the previous block is below the gas target. It repeats until base fee reaches the lower bound fee or the upper bound fee. The base fee per gas is partially burned. Transactions specify the maximum fee per gas they are willing to pay in total. The transaction will always pay the base fee per gas of the block it was included in.
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+## Motivation
+<!--The motivation is critical for KIPs that want to change the Klaytn protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the KIP solves. KIP submissions without sufficient motivation may be rejected outright.-->
+
+Klaytn has been following a policy of “fixed gas price“, with the gas price fixed at a certain level (e.g. 25 ston, 750 ston). This approach was adopted in the initial phase because it allows users to easily send transactions easily without having to enter the gas fee, while minimizing the volatility of transaction fees. But in the periods leading up to the gas price increase, we have seen a lot  ...
```
```suggestion
Klaytn has been following a policy of “fixed gas price“, with the gas price fixed at a certain level (e.g. 25 ston and 750 ston). This approach was adopted in the initial phase because it allows users to easily send transactions without having to predict the gas fee, while minimizing the volatility of transaction fees. But in the periods leading up to the gas price increase, we have seen a lot of transaction bursts on Klaytn with the following problems:
```

lwj010 replied:
```suggestion
Klaytn has been following a policy of “fixed gas price“, with the gas price fixed at a certain level (e.g. 25 ston and 750 ston). This approach was adopted in the initial phase because it allows users to easily send transactions without having to predict the gas fee, while minimizing the volatility of transaction fees. But in the periods leading up to the gas price increase, we have seen a lot of transaction bursts on Klaytn with the following problems:
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+
+Klaytn has been following a policy of “fixed gas price“, with the gas price fixed at a certain level (e.g. 25 ston, 750 ston). This approach was adopted in the initial phase because it allows users to easily send transactions easily without having to enter the gas fee, while minimizing the volatility of transaction fees. But in the periods leading up to the gas price increase, we have seen a lot  ...
+
+* **Delay of transactions** : A large number of transactions were being generated at the same time, leading to increased process time for users. The fundamental cause to this was the sudden surge in transactions. But there were cases of thousands of transactions being fired to process certain transaction have been frequent. And the reason behind this repeated phenomenon was the absence of an algo ...
```
```suggestion
* **Delay of transaction execution** : A large number of transactions were being generated at the same time, leading to increased process time for users. The fundamental cause to this was the sudden surge in transactions. But there were cases of thousands of transactions being fired to process a certain transaction. And the reason behind this repeated phenomenon was the absence of an algorithm that prioritizes Klaytn transactions as well as of an appropriate gas price policy.
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+Klaytn has been following a policy of “fixed gas price“, with the gas price fixed at a certain level (e.g. 25 ston, 750 ston). This approach was adopted in the initial phase because it allows users to easily send transactions easily without having to enter the gas fee, while minimizing the volatility of transaction fees. But in the periods leading up to the gas price increase, we have seen a lot  ...
+
+* **Delay of transactions** : A large number of transactions were being generated at the same time, leading to increased process time for users. The fundamental cause to this was the sudden surge in transactions. But there were cases of thousands of transactions being fired to process certain transaction have been frequent. And the reason behind this repeated phenomenon was the absence of an algo ...
+* **Storage overload** : The aforementioned bot transactions are created mostly by bots, and are reverted transactions. These transactions strain the Klaytn storage in the long-term, which can hinder Klaytn from providing quick transaction finality and stable network.
```
```suggestion
* **Storage overload** : The aforementioned bot transactions are created mostly by bots, and are mostly reverted and useless transactions. These transactions strain the Klaytn storage in the long-term, which can hinder Klaytn from providing quick transaction finality and stable network.
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+
+* **Delay of transactions** : A large number of transactions were being generated at the same time, leading to increased process time for users. The fundamental cause to this was the sudden surge in transactions. But there were cases of thousands of transactions being fired to process certain transaction have been frequent. And the reason behind this repeated phenomenon was the absence of an algo ...
+* **Storage overload** : The aforementioned bot transactions are created mostly by bots, and are reverted transactions. These transactions strain the Klaytn storage in the long-term, which can hinder Klaytn from providing quick transaction finality and stable network.
+* **Inefficient resource allocation due to difficulty in rational price determination** : Gas price should neither be too high or too low. A too low gas price increases vulnerability to DDOS attacks and and transaction delay for users. When it’s too high, the transaction cost is going to prevent users from using the network unreservedly. With the current fixed gas price policy, however, it is dif ...
```
```suggestion
* **Inefficient resource allocation due to difficulty in rational price determination** : Gas price should neither be too high or too low. A too low gas price increases vulnerability to DDOS attacks and transaction delay for users. When it’s too high, the transaction cost is going to prevent users from using the network unreservedly. With the current fixed gas price policy, however, it is difficult to determine which price is appropriate, since the congestion status of the Klaytn network is constantly changing.
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+* **Delay of transactions** : A large number of transactions were being generated at the same time, leading to increased process time for users. The fundamental cause to this was the sudden surge in transactions. But there were cases of thousands of transactions being fired to process certain transaction have been frequent. And the reason behind this repeated phenomenon was the absence of an algo ...
+* **Storage overload** : The aforementioned bot transactions are created mostly by bots, and are reverted transactions. These transactions strain the Klaytn storage in the long-term, which can hinder Klaytn from providing quick transaction finality and stable network.
+* **Inefficient resource allocation due to difficulty in rational price determination** : Gas price should neither be too high or too low. A too low gas price increases vulnerability to DDOS attacks and and transaction delay for users. When it’s too high, the transaction cost is going to prevent users from using the network unreservedly. With the current fixed gas price policy, however, it is dif ...
+* **Inability to respond quickly and flexibly** : Gas price needs to go up during transaction spikes and go down when the network is stable. But under the current fixed gas price scheme, it takes a lot of time until an appropriate price range is discussed and actually implemented.
```
```suggestion
* **Inability to respond quickly and flexibly** : Gas price needs to go up during transaction spikes and go down when the network is not congested. But under the current fixed gas price scheme, it is not easy to analyze and change the gas price based on the network congestion situation.
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+
+A burn mechanism is needed to be introduced on Klaytn while solving the above problems, in order to link the growth of the ecosystem with KLAY value.
+
+(This proposal only deals with dynamic fee policy, and a FIFO (First In First Out) transaction priority algorithm is already built at https://github.com/klaytn/klaytn)
```
```suggestion
(This proposal only deals with dynamic fee policy, and a FIFO (First In First Out) transaction priority algorithm is already implemented. See https://github.com/klaytn/klaytn/pull/1282 and https://github.com/klaytn/klaytn/pull/1309)
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+## Specification
+This specification derived heavily from Ethereum's [ERC-1559](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md) written by Vitalik Buterin (@vbuterin), Eric Conner (@econoar), Rick Dudley (@AFDudley), Matthew Slipper (@mslipper), Ian Norden (@i-norden), Abdelhamid Bakhta (@abdelhamidbakhta).
+
+Transactions under a dynamic gas fee policy consist of `base_fee`, which are dynamically controlled according to the network congestion status. Network congestion is measured by `gas_used` which is the gas usage by blocks created on Klaytn. `base_fee` changes every block.
```
How the `base_fee` is included in the block? Do we use `gasPrice` field for this?

lwj010 replied:
@kjhman21 We already have a `baseFeePerGas` field for that. Refer to https://docs.klaytn.com/klaytn/design/computation/execution-model#block

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+
+Transactions under a dynamic gas fee policy consist of `base_fee`, which are dynamically controlled according to the network congestion status. Network congestion is measured by `gas_used` which is the gas usage by blocks created on Klaytn. `base_fee` changes every block.
+
+When a consensus node creates a block, if the `gas_used` of parent_block exceeds `gas_target`, `base_fee` would go up. On the other hand, if the `gas_used` is lower than `gas_target`, the `base_fee` would be reduced. This process would be repeated until the `base_fee` doesn’t exceed `lower_bound` or `upper_bound`. The block proposer would receive a part of the transaction feeds included in the bl ...
```
```suggestion
When a consensus node creates a block, if the `gas_used` of the parent block exceeds `gas_target`, `base_fee` would go up. On the other hand, if the `gas_used` is lower than `gas_target`, the `base_fee` would be reduced. This process would be repeated until the `base_fee` doesn’t exceed `lower_bound` or `upper_bound`. The block proposer would receive a part of the fee of transactions included in the block, and the rest would be burned.
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+@dataclass
+class TxTypeLegacyTransaction:
+	value: int = 0
+    to: int = 0
```
invalid indentation?

lwj010 replied:
Resolved.

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+			parent_gas_used = self.parent(block).gas_used
+			transactions = self.transactions(block)
+
+			# check if the base fee is in the range
+			if parent_base_fee_per_gas < LOWER_BOUND_BASE_FEE:
+				expected_base_fee_per_gas = LOWER_BOUND_BASE_FEE
+
+			elif parent_base_fee_per_gas > UPPER_BOUND_BASE_FEE:
+				expected_base_fee_per_gas = UPPER_BOUND_BASE_FEE
+
```
Don't we need to first calculate the expected_base_fee_per_gas?
then if the value is not in the range (lower, upper), we need to adjust the value.

lwj010 replied:
I think this work need to be done in line 216~227. Will update it.

lwj010 replied:
@kjhman21 logic updated.

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+					expected_base_fee_per_gas = parent_base_fee_per_gas
+				elif parent_gas_used > GAS_TARGET:
+					gas_used_delta = parent_gas_used - GAS_TARGET
+					base_fee_per_gas_delta = max(parent_base_fee_per_gas * gas_used_delta // GAS_TARGET // BASE_FEE_DELTA_REDUCING_DENOMINATOR, 1)
```
We need to define the operator `//` first.

lwj010 replied:
Added note on line 44

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+## Backwards Compatibility
+
+* This implementation will not break backward compatibility.
+* But to use this feature, you must use another transaction type which will be implemented with this KIP.
```
This is not true. We will use the same transaction type, but the gas price value can vary.

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+
+## Backwards Compatibility
+
+* This implementation will not break backward compatibility.
```
We need to say that this will break backward compatibility I think.
Because setting the gas price value should be changed.

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+					expected_base_fee_per_gas = parent_base_fee_per_gas + base_fee_per_gas_delta
+				else:
+					gas_used_delta = GAS_TARGET - parent_gas_used
+					base_fee_per_gas_delta = parent_base_fee_per_gas * gas_used_delta // GAS_TARGET // BASE_FEE_DELTA_REDUCING_DENOMINATOR
```
Don't we need to make a lower bound using max() like the above?

lwj010 replied:
```suggestion
					base_fee_per_gas_delta = max(parent_base_fee_per_gas * gas_used_delta // GAS_TARGET // BASE_FEE_DELTA_REDUCING_DENOMINATOR, 1)
```

lwj010 committed:
Removed .DS_Store

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 committed:
Removed .DS_Store files

lwj010 committed:
Merge branch 'master' of https://github.com/lwj010/kips

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 committed:
Apply suggestions from code review

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 committed:
Apply suggestions from code review

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+@dataclass
+class TxTypeLegacyTransaction:
+	value: int = 0
+    to: int = 0
```
```suggestion
    value: int = 0
    to: int = 0
```

lwj010 commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+				elif parent_gas_used > GAS_TARGET:
+					gas_used_delta = parent_gas_used - GAS_TARGET
+					base_fee_per_gas_delta = max(parent_base_fee_per_gas * gas_used_delta // GAS_TARGET // BASE_FEE_DELTA_REDUCING_DENOMINATOR, 1)
+					expected_base_fee_per_gas = parent_base_fee_per_gas + base_fee_per_gas_delta
```
```suggestion
					expected_base_fee_per_gas = parent_base_fee_per_gas + base_fee_per_gas_delta
					
					if expected_base_fee_per_gas >  UPPER_BOUND_BASE_FEE:
					    expected_base_fee_per_gas = UPPER_BOUND_BASE_FEE
```

lwj010 commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,381 @@
+				else:
+					gas_used_delta = GAS_TARGET - parent_gas_used
+					base_fee_per_gas_delta = parent_base_fee_per_gas * gas_used_delta // GAS_TARGET // BASE_FEE_DELTA_REDUCING_DENOMINATOR
+					expected_base_fee_per_gas = parent_base_fee_per_gas - base_fee_per_gas_delta
```
```suggestion
					expected_base_fee_per_gas = parent_base_fee_per_gas - base_fee_per_gas_delta
					
					if expected_base_fee_per_gas < LOWER_BOUND_BASE_FEE:
					   expected_base_fee_per_gas = LOWER_BOUND_BASE_FEE 
```

lwj010 committed:
Update kip-71.md

Updated
- LOWER_BOUND and UPPER_BOUND logic
- transaction accounting
- Backwards compatibility

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,387 @@
+		parent_base_fee_per_gas = self.parent(block).base_fee_per_gas
+		transactions = self.transactions(block)
+
+		if parent_gas_used > BLOCK_GAS_LIMIT:
```
Where is this value `parent_gas_used` first assigned? I cannot find the location.

lwj010 replied:
Updated.

kjeom commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,387 @@
+
+INITIAL_FORK_BLOCK_NUMBER = 107806544 # TBD
+BASE_FEE_DELTA_REDUCING_DENOMINATOR = 36 # Max basefee change: 5% per block 
+LOWER_BOUND_BASE_FEE = 2500000000 # 25ston
```
I think you set the LOWER_BOUND_BASE_FEE to 2.5ston

lwj010 replied:
Thanks for your feedback.

```suggestion
LOWER_BOUND_BASE_FEE = 25000000000 # 25ston
```

lwj010 committed:
Update kip-71.md

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: jack <jack.jin@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

lwj010 committed:
Update indentation

lwj010 committed:
Update kip-71.md

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+
+* **Delay of transaction execution** : A large number of transactions were being generated at the same time, leading to increased process time for users. The fundamental cause to this was the sudden surge in transactions. But there were cases of thousands of transactions being fired to process a certain transaction. And the reason behind this repeated phenomenon was the absence of an algorithm th ...
+* **Storage overload** : The aforementioned bot transactions are created mostly by bots, and are mostly reverted and useless transactions. These transactions strain the Klaytn storage in the long-term, which can hinder Klaytn from providing quick transaction finality and stable network.
+* **Inefficient resource allocation due to difficulty in rational price determination** : Gas price should neither be too high or too low. A too low gas price increases vulnerability to DDOS attacks and transaction delay for users. When it’s too high, the transaction cost is going to prevent users from using the network unreservedly. With the current fixed gas price policy, however, it is difficu ...
```
```suggestion
* **Inefficient resource allocation due to difficulty in rational price determination** : Gas price should neither be too high or too low. A too low gas price increases vulnerability to DoS (denial-of-service) attacks and transaction delay for users. When it’s too high, the transaction cost is going to prevent users from using the network unreservedly. With the current fixed gas price policy, however, it is difficult to determine which price is appropriate, since the congestion status of the Klaytn network is constantly changing.
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+(This proposal only deals with dynamic fee policy, and a FIFO (First In First Out) transaction priority algorithm is already implemented. See https://github.com/klaytn/klaytn/pull/1282 and https://github.com/klaytn/klaytn/pull/1309)
+
+## Specification
+This specification derived heavily from Ethereum's [ERC-1559](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md) written by Vitalik Buterin (@vbuterin), Eric Conner (@econoar), Rick Dudley (@AFDudley), Matthew Slipper (@mslipper), Ian Norden (@i-norden), Abdelhamid Bakhta (@abdelhamidbakhta).
```
```suggestion
This specification derived heavily from Ethereum's [ERC-1559](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md) written by Vitalik Buterin (@vbuterin), Eric Conner (@econoar), Rick Dudley (@AFDudley), Matthew Slipper (@mslipper), Ian Norden (@i-norden), and Abdelhamid Bakhta (@abdelhamidbakhta).
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+## Specification
+This specification derived heavily from Ethereum's [ERC-1559](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md) written by Vitalik Buterin (@vbuterin), Eric Conner (@econoar), Rick Dudley (@AFDudley), Matthew Slipper (@mslipper), Ian Norden (@i-norden), Abdelhamid Bakhta (@abdelhamidbakhta).
+
+Transactions under a dynamic gas fee policy consist of `base_fee`, which are dynamically controlled according to the network congestion status. Network congestion is measured by `gas_used` which is the gas usage by blocks created on Klaytn. `base_fee` changes every block.
```
@lwj010 If you use `base_fee` like this, it means there is a field with that name, but it seems not true. use `baseFeePerGas` or just use base fee?

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+INITIAL_FORK_BLOCK_NUMBER = 107806544 # TBD
+BASE_FEE_DELTA_REDUCING_DENOMINATOR = 36 # Max basefee change: 5% per block 
+LOWER_BOUND_BASE_FEE = 25000000000 # 25ston
+UPPER_BOUND_BASE_FEE = 750000000000 # 750ston
```
```suggestion
UPPER_BOUND_BASE_FEE = 750000000000 # 750 ston
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+LOWER_BOUND_BASE_FEE = 25000000000 # 25ston
+UPPER_BOUND_BASE_FEE = 750000000000 # 750ston
+GAS_TARGET = 30000000 
+BLOCK_GAS_LIMIT = 84000000 # implicit block gas limit to enforce the max basefee change rate
```
It seems that others can mislead Klaytn have block gas limit in the future. I think we can choose any other name for this. `MAX_BLOCK_GAS_USED_FOR_BASE_FEE`?

kjeom replied:
I also think the name, `BLOCK_GAS_LIMIT` can be confusing in ethereum's one. We use it only for calculating `BASE_FEE`

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+UPPER_BOUND_BASE_FEE = 750000000000 # 750ston
+GAS_TARGET = 30000000 
+BLOCK_GAS_LIMIT = 84000000 # implicit block gas limit to enforce the max basefee change rate
+BURN_RATIO = 0.5 
```
We make sure distinguish which values are for governance parameters or not.

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+			if expected_base_fee_per_gas < LOWER_BOUND_BASE_FEE:
+				expected_base_fee_per_gas = LOWER_BOUND_BASE_FEE
+
+		assert expected_base_fee_per_gas == block.base_fee_per_gas, 'invalid block: base fee not correct'
```
@lwj010 Why do we have this assert?

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+
+			
+			signer.balance -= transaction.amount
+			assert signer.balance >= 0, 'invalid transaction: signer does not have enough ETH to cover attached value'
```
```suggestion
			assert signer.balance >= 0, 'invalid transaction: signer does not have enough KLAY to cover attached value'
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+			
+			# TODO: fee delegation transaction accounting
+			# the signer must be able to afford the transaction
+			assert signer.balance >= transaction.gas_limit * transaction.max_fee_per_gas
```
```suggestion
			assert fee_payer.balance >= transaction.gas_limit * transaction.max_fee_per_gas
```

we can replace signer to fee_payer so that we can include fee delegation transactions. i.e., we can think of a sender of non-fee-delegated transactions as a fee-payer.

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+			
+
+			signer.balance -= transaction.gas_limit * block.base_fee_per_gas
+			assert signer.balance >= 0, 'invalid transaction: signer does not have enough ETH to cover gas'
```
```suggestion
			assert fee_payer.balance >= 0, 'invalid transaction: signer does not have enough KLAY to cover gas'
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+			assert transaction.max_fee_per_gas < 2**256
+			
+
+			signer.balance -= transaction.gas_limit * block.base_fee_per_gas
```
```suggestion
			fee_payer.balance -= transaction.gas_limit * block.base_fee_per_gas
```

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+			gas_refund = transaction.gas_limit - gas_used
+			cumulative_transaction_gas_used += gas_used
+			# signer gets refunded for unused gas
+			signer.balance += gas_refund * block.base_fee_per_gas
```
```suggestion
			fee_payer.balance += gas_refund * block.base_fee_per_gas
```

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 committed:
Update kip-71.md

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+		# check if the block spent too much gas transactions
+		assert cumulative_transaction_gas_used == block.gas_used, 'invalid block: gas_used does not equal total gas used in all transactions'
+
+		# TODO: verify account balances match block's account balances (via state root comparison)
```
@lwj010 Does this TODO need to be resolved? or will you leave this as it is?

lwj010 replied:
@kjhman21 will leave this as it is, since now we're in implementation phase already.

baerae-colin replied:
@lwj010 Do you want us to update this? I think it is now ready to be merged, but it seems not good to have 'todo' here.
cc @kjeom

lwj010 replied:
@kjhman21 removed all todos from the file.

lwj010 commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+
+		# check if the block spent too much gas transactions
+		assert cumulative_transaction_gas_used == block.gas_used, 'invalid block: gas_used does not equal total gas used in all transactions'
+
+		# TODO: verify account balances match block's account balances (via state root comparison)
```
```suggestion


```

lwj010 commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+
+		# TODO: verify account balances match block's account balances (via state root comparison)
+		# TODO: validate the rest of the block
+
```
```suggestion
		# validate the rest of the block

```

lwj010 commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+			signer.balance -= transaction.amount
+			assert signer.balance >= 0, 'invalid transaction: signer does not have enough KLAY to cover attached value'
+			
+			# TODO: fee delegation transaction accounting
```
```suggestion
			# fee delegation transaction accounting is needed 
```

lwj010 commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+
+Transaction = Union[TxTypeLegacyTransaction, TxTypeFeeDelegatedSmartContractExecution, TxTypeFeeDelegatedSmartContractExecutionWithRatio, EthereumTransactions]
+
+# TODO: transaction accounting part
```
```suggestion

```

lwj010 committed:
Update KIPs/kip-71.md

lwj010 committed:
Update KIPs/kip-71.md

lwj010 committed:
Update KIPs/kip-71.md

lwj010 committed:
Update KIPs/kip-71.md

lwj010 reviewed (commented):
Removed todos from the document.

baerae-colin commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,385 @@
+Transactions under a dynamic gas fee policy consist of `base_fee_per_gas`, which are dynamically controlled according to the network congestion status. Network congestion is measured by `gas_used` which is the gas usage by blocks created on Klaytn. `base_fee_per_gas` changes every block.
+
+When a consensus node creates a block, if the `gas_used` of the parent block exceeds `gas_target`, `base_fee_per_gas` would go up. On the other hand, if the `gas_used` is lower than `gas_target`, the `base_fee_per_gas` would be reduced. This process would be repeated until the `base_fee_per_gas` doesn’t exceed `lower_bound` or `upper_bound`. The block proposer would receive a part of the fee of t ...
+
```
```suggestion

The below shows pseudo code of this proposal.

```

baerae-colin reviewed (approved):
@kjeom Please take a look. I'm okay with it. Thanks @lwj010

kjeom commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,385 @@
+
+* All previous transaction types will remain the same.
+* Setting fixed gas price to send transactions will still work and be included in blocks, but those transactions will be not included in the block if a `base_fee` is higher than the gas price.
+
```
I suggest to replace `base_fee` to `base_fee_per_gas` to avoid confusion.
Furthermore I'd like to emphasize `base_fee_per_gas` in txPool and `gas_price` in tx.

kjeom replied:
Everything else LGTM.

lwj010 replied:
renamed it to `base_fee_per_gas`.

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Junghyun Colin Kim <colin.klaytn@krustuniverse.com>

lwj010 commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,385 @@
+## Backwards Compatibility
+
+* All previous transaction types will remain the same.
+* Setting fixed gas price to send transactions will still work and be included in blocks, but those transactions will be not included in the block if a `base_fee` is higher than the gas price.
```
```suggestion
* Setting fixed gas price to send transactions will still work and be included in blocks, but those transactions will be not included in the block if a `base_fee_per_gas` is higher than the gas price.
```

lwj010 reviewed (commented):
`base_fee` naming to `base_fee_per_gas`

lwj010 committed:
Update KIPs/kip-71.md

lwj010 commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+## Backwards Compatibility
+
+* All previous transaction types will remain the same.
+* Setting fixed gas price to send transactions will still work and be included in blocks, but those transactions will be not included in the block if a `base_fee` is higher than the gas price.
```
```suggestion
* Setting fixed gas price to send transactions will still work and be included in blocks, but those transactions will be not included in the block if a `base_fee_per_gas` is higher than the gas price.
```

aidan-kwon commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+Klaytn has been following a policy of “fixed gas price“, with the gas price fixed at a certain level (e.g. 25 ston and 750 ston). This approach was adopted in the initial phase because it allows users to easily send transactions without having to predict the gas fee, while minimizing the volatility of transaction fees. But in the periods leading up to the gas price increase, we have seen a lot of ...
+
+* **Delay of transaction execution** : A large number of transactions were being generated at the same time, leading to increased process time for users. The fundamental cause to this was the sudden surge in transactions. But there were cases of thousands of transactions being fired to process a certain transaction. And the reason behind this repeated phenomenon was the absence of an algorithm th ...
+* **Storage overload** : The aforementioned bot transactions are created mostly by bots, and are mostly reverted and useless transactions. These transactions strain the Klaytn storage in the long-term, which can hinder Klaytn from providing quick transaction finality and stable network.
```
`bot transactions were created by bot` is weird for me  
```suggestion
* **Storage overload** : The aforementioned a large number of transactions are created mostly by bots, and are mostly reverted and useless transactions. These transactions strain the Klaytn storage in the long-term, which can hinder Klaytn from providing quick transaction finality and stable network.
```

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Donghwan Aidan Kwon <aidan.kwon@krustuniverse.com>

aidan-kwon commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+
+			# Prevent impossibly large numbers
+			assert transaction.max_fee_per_gas < 2**256
+			
```
```suggestion
```

aidan-kwon commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+
+			signer = self.account(signer_address)
+
+			
```
```suggestion
```

aidan-kwon commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+			gas_used = self.execute_transaction(transaction, block.base_fee_per_gas)
+			gas_refund = transaction.gas_limit - gas_used
+			cumulative_transaction_gas_used += gas_used
+			# signer gets refunded for unused gas
```
```suggestion

			# signer gets refunded for unused gas
```

aidan-kwon commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+			# Note: this validates transaction signature and chain ID which must happen before we normalize below since normalized transactions don't include signature or chain ID
+			signer_address = self.validate_and_recover_signer_address(unnormalized_transaction)
+			transaction = self.normalize_transaction(unnormalized_transaction, signer_address)
+
```
```suggestion
```

aidan-kwon commented on KIPs/kip-71.md:
```diff
@@ -0,0 +1,386 @@
+
+		# check if the block spent too much gas transactions
+		assert cumulative_transaction_gas_used == block.gas_used, 'invalid block: gas_used does not equal total gas used in all transactions'
+
```
```suggestion
```

aidan-kwon reviewed (approved):
LGTM. Thanks!

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Donghwan Aidan Kwon <aidan.kwon@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Donghwan Aidan Kwon <aidan.kwon@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Donghwan Aidan Kwon <aidan.kwon@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Donghwan Aidan Kwon <aidan.kwon@krustuniverse.com>

lwj010 committed:
Update KIPs/kip-71.md

Co-authored-by: Donghwan Aidan Kwon <aidan.kwon@krustuniverse.com>

baerae-colin commented:
@lwj010 hmm... it seems something went wrong. Could you push another commit to run CI once again?

lwj010 reviewed (commented):
Added extra review

lwj010 committed:
Update kip-71.md

lwj010 commented:
> @lwj010 hmm... it seems something went wrong. Could you push another commit to run CI once again?

Just pushed another commit.
